### ISSUE TITLE
Spike in get well known endpoints in v4

### ISSUE THREAD

**Issue by garrettmac** [okta/okta-auth-js#654]
When updating from "^3.1.4" to "^4.7.1" we noticed 2x jump in calls to our "^3.1.4". I may have to downgrade to your early version :(

Can you provide insight in why/when this is getting called for me to better understand?

For me this is the order im making calls if it helps
1. authClient.signIn
2. authClient.token.getWithoutPrompt
3. authClient.tokenManager.add("accessToken"
4. authClient.tokenManager.add("idToken",

**Comment by shuowu** [comment]
@garrettmac It might be related to the `active auto renew` behavior that has been introduced since [v4.0](https://github.com/okta/okta-auth-js/blob/master/CHANGELOG.md#breaking-changes). It calls the token endpoint automatically when they are about to expire. 

You can disable the autoRenew feature with the [authRenew config](https://github.com/okta/okta-auth-js#autorenew) if you prefer to manage the token renew by yourself.

**Comment by garrettmac** [comment]
@shuowu thanks for the speedy reply 

ah would that also be related to why i see this fire twice?

<img width="618" alt="Screen Shot 2021-03-09 at 1 46 10 PM" src="https://user-images.githubusercontent.com/5916111/110542454-e8fdb180-80dd-11eb-8ace-3f6532caf83d.png">

I dont have any subscription on my end so assume so for auto renew.

So what if i want auto renew? why would you be calling for that to renew for this login flow if its a new session?

(so which if any of these functions above would call that) 
I still want the auto renew feature, just dont need it called 2x for logging in?.

**Comment by shuowu** [comment]
Per the warning message, you can check it in #655.

> So what if i want auto renew? why would you be calling for that to renew for this login flow if its a new session?

Do you mean the sdk called some endpoint 2x times during the login process? Can you share the network log, then I can take a deeper look into it.

**Comment by garrettmac** [comment]
Yeah @shuowu  I noticed x2 the amount of calls to the well-known endpoint in system logs but repo'ing I dont see this being the case in the network tab. So maybe some more investigation on my end.
Can you elaborate on when the well-known endpoint gets called? is it only when `okta-cache-storage` is missing (so gets fired on initiation)? or when it makes updates to the renew token or other places too?

**Comment by shuowu** [comment]
@garrettmac With some digging:
1) The get well-known call is cached with `okta-cache-storage` https://github.com/okta/okta-auth-js/blob/6d16f8f98665d9d1034afe3aa14dbf2af24c3978/lib/oidc/endpoints/well-known.ts#L20-L22
2) The expire time has been set as 86400s (24h) as default. https://github.com/okta/okta-auth-js/blob/236a4a4e54e634a36e004be93a4cda2580198dd2/lib/http.ts#L76-L81

Looks like the logic above should be able to avoid unnecessary request to the endpoint.

**Comment by garrettmac** [comment]
@shuowu 
That I observed as well. If:
#### step 1.
  user does login => (**Network Call to .well-known/openid-configuration** fires because `okta-cache-storage` is missing, now it has it, its cached)
#### step 2. 
user does logout  => `okta-cache-storage` remains, no impact
#### step 3.
 user logs in again => (**NO Network Call to .well-known/openid-configuration**, reads from `okta-cache-storage` in caches)

<br>
That I understand. Seems the same with both v3 and v4.

### So still not of some of the basics
1. With my config in v3 (or v4 for that matter) i have auto renew on by default correct? since im missing `autoRenew: false`, correct?
2. When does .well-known/openid-configuration get called again if `okta-cache-storage` is set? on expiry? any other condition?
3. Why would you suggest auto renewal could be why? How are .well-known/openid-configuration calls impacted by auto renewal on version 3 vs 4 
4. May this will be inherently answered in the above, in v4 why would I see 2x the .well-known/openid-configuration calls with v4 compared to v3 if step 1-3 happen the same with both v3 and v4? Shouldn't they be the same in those regards?
5. I noticed you provide the version number in a header, would anything else have changed that could trip me up from a reporting perspective if it is in fact an issue with my reporting when on v4. Doesnt seem like i see more network calls to .well-known/openid-configuration as a guest going through the flow as a guest.


here is my config btw:
```js
 {
        url: "https://xxxxxxx.okta.com",
        clientId: "xxxxxxxxx",
        issuer: "https://xxxxxxxx.okta.com/oauth2/default",
        pkce: true,
        tokenManager: { secure: true }
      }
```

**Comment by garrettmac** [comment]
1. https://github.com/okta/okta-auth-js/tree/okta-auth-js-3.1.4#configuration-options

> **By default, the tokenManager will attempt to renew expired tokens.** When an expired token is requested by the tokenManager.get() method, a renewal request is executed to update the token. If you wish to manually control token renewal, **set autoRenew to false to disable this feature.** 

I think 1, is clear, just would like confirmation from you @shuowu

**Comment by shuowu** [comment]
@garrettmac 
> Why would you suggest auto renewal could be why? How are .well-known/openid-configuration calls impacted by auto renewal on version 3 vs 4

Sorry to confuse you about the active auto renew change in v4, I miss read the issue at the first glance (missed the well known endpoint, I thought you were asking about token requests..,). The behaviour should be consistent across versions. 

Also, if you see two `openid-configuration` requests in your network log, the first one is a [preflight request](https://developer.mozilla.org/en-US/docs/Glossary/Preflight_request) for cors purpose.

**Comment by garrettmac** [comment]
yeah preflight is normal for all calls, was able to get a snapshot. Its odd that when it calls openid-configuration before the authn call we get this, not when authn fires first (and it normally looks like it does) so im thinking a race condition? 
<img width="564" alt="Screen Shot 2021-03-11 at 4 38 48 PM" src="https://user-images.githubusercontent.com/5916111/110874360-741aab00-8288-11eb-9ca2-d246587c8e7f.png">

**Comment by shuowu** [comment]
@garrettmac 
authClient.signIn basically calls `signInWithCredentials`, which should not trigger request the well know endpoint.
https://github.com/okta/okta-auth-js/blob/5e9af2fdc6d3b4925e2981b3e34529c86aa2fa3e/lib/browser/browser.ts#L274-L291

Normally, the well-known endpoint should be requested in your second step (authClient.token.getWithoutPrompt). As you mentioned, it looks like a race condition. I suggest reaching out to our support team at developers@okta.com to cover the reproduction details of your issue.

**Comment by garrettmac** [comment]
ok I reached out... (why email though? why not here?)
they said they couldnt repo ugg. Hoping 4.8.0 resolves the issue 
<img width="921" alt="Screen Shot 2021-03-18 at 9 41 27 AM" src="https://user-images.githubusercontent.com/5916111/111686531-69ab6480-87e6-11eb-9b29-73b8dced1fdf.png">

**Comment by garrettmac** [comment]
nope, but seems the extra open id calls are not not being made before the user click action, instead in the background, in what i can only guess is the 24hr session expiry, (its reads from okta-cache-storage in caches and it decides to update here as you provided in the breaking changes: 
<img width="500" alt="Screen Shot 2021-03-19 at 10 24 43 AM" src="https://user-images.githubusercontent.com/5916111/111824093-f74c8a00-88a2-11eb-98f8-2d2325bc4702.png">
 
seems autoRenew: false is all i need. thanks

**Comment by shuowu** [comment]
Right, it's about the active renew scenario. But I don't think it will cause the well-known endpoint request spike.

**Comment by garrettmac** [comment]
so found a way to repo.
edit  "okta-cache-storage" and "okta-token-storage" to have expired expiresAt then look at your network. it fires the well known a ton of times

**Comment by aarongranick-okta** [comment]
@garrettmac Thanks for providing a snippet. We will take a look and hopefully this will help us reproduce and identify the issue that you are seeing. internal ref: OKTA-380363

**Comment by garrettmac** [comment]
no impact on if ` autoRenew: false,` is there, doesnt seem i can opt out, there is a listener i dont want. 
the token and cache (openid) calls fire 3x each (6 if you count the preflight)

**Comment by garrettmac** [comment]
Lines fired from the lib causing this is https://github.com/okta/okta-auth-js/blob/master/lib/TokenManager.ts#L485L497. autoRemove: false seems to make reverse compatible

**Comment by garrettmac** [comment]
i see when it was introduced in 4.1.0 was set to true by default https://github.com/okta/okta-auth-js/pull/484/files#diff-f18cf5684cf053b7ffbc2723f4983f1a26432ef9eade50cbfa00a74a7793dedbR33

**Comment by garrettmac** [comment]
@aarongranick-okta could yall repo? (i see 3x as many calls for 3 endpoints. keys, openid and token)
